ddtrace/tracer/handler.go

```diff
@@ -0,0 +1,235 @@
+	flush()
+
+	// stop shuts down the traceWriter, ensuring all payloads are flushed.
+	stop()
+}
+
+type agentTraceWriter struct {
+	config *config
+
+	payload *payload
+
+	// climit limits the number of concurrent outgoing connections
+	climit chan struct{}
+
+	// wg waits for all goroutines to exit when stopping.
+	wg sync.WaitGroup
+
+	// prioritySampling refers to the tracer's priority sampler.
+	prioritySampling *prioritySampler
+}
+
+var _ traceWriter = &agentTraceWriter{}
+
+func newAgentTraceWriter(c *config, s *prioritySampler) *agentTraceWriter {
+	return &agentTraceWriter{
+		config:           c,
+		payload:          newPayload(),
+		climit:           make(chan struct{}, concurrentConnectionLimit),
+		prioritySampling: s,
+	}
+}
+
+func (h *agentTraceWriter) add(trace []*span) {
+	if err := h.payload.push(trace); err != nil {
+		h.config.statsd.Incr("datadog.tracer.traces_dropped", []string{"reason:encoding_error"}, 1)
+		log.Error("error encoding msgpack: %v", err)
+	}
+	if h.payload.size() > payloadSizeLimit {
+		h.config.statsd.Incr("datadog.tracer.flush_triggered", []string{"reason:size"}, 1)
+		h.flush()
+	}
+}
+
+func (h *agentTraceWriter) stop() {
+	h.wg.Wait()
+}
+
+// flush will push any currently buffered traces to the server.
+func (h *agentTraceWriter) flush() {
+	if h.payload.itemCount() == 0 {
+		return
+	}
+	h.wg.Add(1)
+	h.climit <- struct{}{}
+	go func(p *payload) {
+		defer func(start time.Time) {
+			<-h.climit
+			h.wg.Done()
+			h.config.statsd.Timing("datadog.tracer.flush_duration", time.Since(start), nil, 1)
+		}(time.Now())
+		size, count := p.size(), p.itemCount()
+		log.Debug("Sending payload: size: %d traces: %d\n", size, count)
+		rc, err := h.config.transport.send(p)
+		if err != nil {
+			h.config.statsd.Count("datadog.tracer.traces_dropped", int64(count), []string{"reason:send_failed"}, 1)
+			log.Error("lost %d traces: %v", count, err)
+		} else {
+			h.config.statsd.Count("datadog.tracer.flush_bytes", int64(size), nil, 1)
+			h.config.statsd.Count("datadog.tracer.flush_traces", int64(count), nil, 1)
+			if err := h.prioritySampling.readRatesJSON(rc); err != nil {
+				h.config.statsd.Incr("datadog.tracer.decode_error", nil, 1)
+			}
+		}
+	}(h.payload)
+	h.payload = newPayload()
+}
+
+type hexInt uint64
+
+func (i hexInt) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + strconv.FormatUint(uint64(i), 16) + `"`), nil
+}
+
+type jsonSpan struct {
+	TraceID  hexInt             `json:"trace_id"`
+	SpanID   hexInt             `json:"span_id"`
+	ParentID hexInt             `json:"parent_id"`
+	Name     string             `json:"name"`
+	Resource string             `json:"resource"`
+	Error    int32              `json:"error"`
+	Meta     map[string]string  `json:"meta"`
+	Metrics  map[string]float64 `json:"metrics"`
+	Start    int64              `json:"start"`
+	Duration int64              `json:"duration"`
+	Service  string             `json:"service"`
+}
+
+type lambdaTraceWriter struct {
```

Comment:
```suggestion
type jsonTraceWriter struct {
```
Could be clearer like this?